### PR TITLE
User defined metadata flags

### PR DESCRIPTION
**Pull request by AndrewGaspar**
See #21

### CODE REVIEW

File: src/interface/Metadata.hpp [hunk 1]
```diff
@@ -76,218 +158,174 @@ class Metadata {
...
+    if (bit.flag_ >= bits_.size()) {
+      bits_.resize(bit.flag_ + 1);
     }
+    bits_[bit.flag_] = flag;
   }
 
 
   /// Checks if the bit is a topology bit
-  bool _isTopology(flags bit) const {
+  bool IsTopology(MetadataFlag bit) const {
```
[the remark this answers is not in the document]
Comment (jlippuner) [reply src/interface/Metadata.hpp]: I'm fine with doing this in a separate PR

File: src/interface/Container.cpp [hunk 2]
```diff
@@ -63,39 +63,47 @@ void Container<T>::Add(const std::string label,
   std::array<int, 6> arrDims;
   calcArrDims_(arrDims, dims);
   // branch on kind of variable
-  if (metadata.hasSparse()) {
+  if (metadata.IsSet(Metadata::Sparse)) {
     // add a sparse variable
     s->_sparseVars.Add(*pmy_block, label, metadata, dims);
-  } else if ( metadata.where() == (Metadata::edge) ) {
+  } else if ( metadata.Where() == (Metadata::Face) ) {
```
[the remark this answers is not in the document]
Comment (jdolence) [reply src/interface/Container.cpp]: I wouldn't worry about this.  I'm messing with such things in #75

**Comment by AndrewGaspar** [comment]
I removed it because it wasn't being built anyway and had long drifted from changes we've made.

File: src/interface/Container.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndrewGaspar) [reply src/interface/Container.cpp]: Yeah, wasn't focused on making style changes - I don't think this should be a blocker.

File: src/interface/Metadata.hpp [hunk 3]
```diff
@@ -13,22 +13,79 @@
...
-///
-///  Can set or query attributes specifed in flags.
-///
+
+namespace internal {
+  enum class MetadataInternal {
+    // declare all the internal flags in an enum so that their values are unique and kept up to date
+#define PARTHENON_INTERNAL_FOR_FLAG(name) name,
+    PARTHENON_INTERNAL_FOREACH_BUILTIN_FLAG
+    Max
```
[the remark this answers is not in the document]
Comment (AndrewGaspar) [reply src/interface/Metadata.hpp]: You can't - there's no `Metadata::Max`.

File: src/interface/ContainerIterator.hpp [hunk 4]
```diff
@@ -107,21 +86,21 @@ class ContainerIterator {
...
-  static bool couldBeFace(const std::vector<Metadata::flags> &flagVector) {
+  static bool couldBeFace(const std::vector<MetadataFlag> &flagVector) {
     // returns true if face is set or if no topology set
     for (auto &f : flagVector) {
-      if ( f == Metadata::face) return true;
-      else if ( f == Metadata::cell) return false;
-      else if ( f == Metadata::edge) return false;
+      if ( f == Metadata::Face) return true;
+      else if ( f == Metadata::Cell) return false;
+      else if ( f == Metadata::Edge) return false;
```
[the remark this answers is not in the document]
Comment (AndrewGaspar) [reply src/interface/ContainerIterator.hpp]: If this needs to be changed, let's do it in another PR

[the remark this answers is not in the document]
Comment (jlippuner) [reply src/interface/ContainerIterator.hpp]: But it's just adding `else if ( f == Metadata::Node) return false;` in both functions, isn't it? That seems easy enough to include in this PR

File: src/interface/Metadata.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Yurlungur) [reply src/interface/Metadata.hpp]: PR created. #76. @jlippuner do you want to take a look when you get the chance? (Already cleared with @jdolence )

**Comment by AndrewGaspar** [comment]
The rename thing was just because the files look similar due to the license at the top. Git figures out renames via fuzzy match.

**Pull request merged**